radixdlt/src/main/java/com/radixdlt/consensus/VertexStore.java

```diff
@@ -115,73 +182,164 @@ public void processGetVerticesRequest(GetVerticesRequest request) {
+		// TODO: cleanup syncs which have views lower than this
+		syncState.fetched.sort(Comparator.comparing(Vertex::getView));
+		List<Vertex> nonRootVertices = syncState.fetched.stream().skip(1).collect(Collectors.toList());
+		rebuild(syncState.fetched.get(0), syncState.fetched.get(1).getQC(), nonRootVertices);
+
+		this.syncToQC(syncState.qc, syncState.committedQC, syncState.author);
+	}
+
+	public void processCommittedStateSync(CommittedStateSync committedStateSync) {
+		log.info("SYNC_STATE: synced {}", committedStateSync);
+
+		final Hash syncTo = (Hash) committedStateSync.getOpaque();
 		SyncState syncState = syncing.get(syncTo);
-		if (syncState == null) {
-			return;
+		if (syncState != null) {
+			rebuildAndSyncQC(syncState);
 		}
+	}
 
-		final QuorumCertificate qc = syncState.qc;
+	private void processVerticesResponseForCommittedSync(Hash syncTo, SyncState syncState, GetVerticesResponse response) {
+		log.info("SYNC_STATE: Processing vertices {}", syncState);
 
-		if (response.getVertices().isEmpty()) {
-			log.info("GET_VERTICES failed: {}", qc);
-			// failed
-			return;
+		long stateVersion = syncState.committedVertexMetadata.getStateVersion();
+		List<ECPublicKey> signers = Collections.singletonList(syncState.author);
+		syncState.fetched.addAll(response.getVertices());
+
+		if (syncedStateComputer.syncTo(stateVersion, signers, syncTo)) {
+			rebuildAndSyncQC(syncState);
+		} else {
+			syncState.syncStage = SyncStage.SYNC_TO_COMMIT;
 		}
+	}
 
+	private void processVerticesResponseForQCSync(Hash syncTo, SyncState syncState, GetVerticesResponse response) {
 		Vertex vertex = response.getVertices().get(0);
 		syncState.fetched.addFirst(vertex);
 		Hash nextVertexId = vertex.getQC().getProposed().getId();
 
 		if (vertices.containsKey(nextVertexId)) {
 			for (Vertex v: syncState.fetched) {
 				if (!addQC(v.getQC())) {
-					log.info("GET_VERTICES failed: {}", qc);
+					log.info("GET_VERTICES failed: {}", syncState.qc);
 					return;
 				}
 				try {
-					insertVertex(v);
+					insertVertexInternal(v);
 				} catch (VertexInsertionException e) {
 					log.info("GET_VERTICES failed: {}", e.getMessage());
 					return;
 				}
 			}
-			addQC(qc);
+			addQC(syncState.qc);
 		} else {
-			log.info("SYNC_VERTICES: Sending further GetVerticesRequest {}", nextVertexId);
+			log.info("SYNC_VERTICES: Sending further GetVerticesRequest for qc={} fetched={} root={}",
+				syncState.qc, syncState.fetched.size(), vertices.get(rootId.get()));
 			syncVerticesRPCSender.sendGetVerticesRequest(nextVertexId, syncState.author, 1, syncTo);
 		}
 	}
 
-	private void doSync(QuorumCertificate qc, ECPublicKey author) {
-		final Hash vertexId = qc.getProposed().getId();
+	public void processGetVerticesResponse(GetVerticesResponse response) {
+		log.info("SYNC_VERTICES: Received GetVerticesResponse {}", response);
+
+		final Hash syncTo = (Hash) response.getOpaque();
+		SyncState syncState = syncing.get(syncTo);
+		if (syncState == null) {
+			return;
```

Comment:
no shouldn't be a malicious response since the RPC layer should be handling matching requests/responses. This signals that the sync was handled by a different sync and was already processed.